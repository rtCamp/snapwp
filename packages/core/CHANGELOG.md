# @snapwp/core

## 0.2.1

### Patch Changes

-   [#108](https://github.com/rtCamp/snapwp/pull/108) [`c1ee556`](https://github.com/rtCamp/snapwp/commit/c1ee5564e1046411dac0ccb5ace99c59fe337959) Thanks [@Pathan-Amaankhan](https://github.com/Pathan-Amaankhan)! - chore: Enforce "import/default" eslint ruleset.

## 0.2.0

### Minor Changes

-   [#103](https://github.com/rtCamp/snapwp/pull/103) [`579ba54`](https://github.com/rtCamp/snapwp/commit/579ba54e72d78b8b5fae89c8ddcd6bbdc0487f91) Thanks [@ashutoshgautams](https://github.com/ashutoshgautams)! - refactor!: Change `location` to `groupLocation` for script placement enum type

### Patch Changes

-   [#42](https://github.com/rtCamp/snapwp/pull/42) [`dcecabf`](https://github.com/rtCamp/snapwp/commit/dcecabfa9df535727e988d2db59bb0a6aa5d2a73) Thanks [@Swanand01](https://github.com/Swanand01)! - feat: Add support for providing block definitions via the Config API.

-   [#87](https://github.com/rtCamp/snapwp/pull/87) [`71bd814`](https://github.com/rtCamp/snapwp/commit/71bd81452935736170e81dabe8fea48b6d2b8085) Thanks [@BhumikP](https://github.com/BhumikP)! - dev: Implement noPropertyAccessFromIndexSignature rule in tsconfig base

## 0.1.0

### Minor Changes

-   [#81](https://github.com/rtCamp/snapwp/pull/81) [`2595e37`](https://github.com/rtCamp/snapwp/commit/2595e376efb9a24b9caa0be9146976ec1386ffc4) Thanks [@justlevine](https://github.com/justlevine)! - chore: SemVer bump to 0.1.0

### Patch Changes

-   [#52](https://github.com/rtCamp/snapwp/pull/52) [`47cad60`](https://github.com/rtCamp/snapwp/commit/47cad6075621da9946a29feba62fc33fe59fdaf6) Thanks [@Swanand01](https://github.com/Swanand01)! - feat: Add better type definitions for block props

## 0.0.2

### Minor Changes

-   [#62](https://github.com/rtCamp/snapwp/pull/62) [`55bd683`](https://github.com/rtCamp/snapwp/commit/55bd683e11c556bb78140299554cf845ba34903c) Thanks [@BhumikP](https://github.com/BhumikP)! - feat: Add support for overloading html-react-parser options

### Patch Changes

-   [#74](https://github.com/rtCamp/snapwp/pull/74) [`f655e18`](https://github.com/rtCamp/snapwp/commit/f655e18f08f0f1c2402f8a79eb618096346dead5) Thanks [@Swanand01](https://github.com/Swanand01)! - fix: Ensure `homeUrl` and `nextUrl` do not have a trailing slash in `SnapWPConfigManager.normalizeConfig`

-   docs: Update, backfill, and cleanup readmes
-   refactor!: Update `getConfig()` to source separate `SnapWPEnv` and `SnapWPConfig`, and add support for `snapwp.config.ts`
-   chore: use explicit `type` import syntax
-   chore: Cleanup tsconfig rulesets and remediate unearthed issues
-   chore: remediate `noUncheckedIndexedAccess` violations
