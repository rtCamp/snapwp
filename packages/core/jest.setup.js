global.__snapWPConfig = {};

global.__envConfig = {
	frontendUrl: 'https://env-next.example.com',
	wpHomeUrl: 'https://env-home.example.com',
	wpSiteUrl: 'https://env-home.example.com/wp',
	graphqlEndpoint: 'env-index.php?graphql',
	uploadsDirectory: '/env-wp-content/uploads',
	restUrlPrefix: '/env-wp-json',
};

process.env.NEXT_PUBLIC_FRONTEND_URL = global.__envConfig.frontendUrl;
process.env.NEXT_PUBLIC_WP_HOME_URL = global.__envConfig.wpHomeUrl;
process.env.WP_SITE_URL = global.__envConfig.wpSiteUrl;
process.env.GRAPHQL_ENDPOINT = global.__envConfig.graphqlEndpoint;
process.env.WP_UPLOADS_DIRECTORY = global.__envConfig.uploadsDirectory;
process.env.REST_URL_PREFIX = global.__envConfig.restUrlPrefix;
