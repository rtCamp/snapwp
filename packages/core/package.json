{
	"name": "@snapwp/core",
	"version": "0.0.2",
	"license": "AGPL-3.0",
	"description": "Shared utility methods and React components used by our other packages.",
	"author": "rtCamp",
	"type": "module",
	"engines": {
		"node": ">=20.0.0",
		"npm": ">=10.0.0",
		"yarn": "please-use-npm"
	},
	"files": [
		"dist"
	],
	"exports": {
		".": {
			"import": "./dist/index.js",
			"require": "./dist/index.js",
			"types": "./dist/types/index.d.ts"
		},
		"./parseExternalRemotePatterns": {
			"import": "./dist/utils/parseExternalRemotePatterns.js"
		},
		"./config": {
			"import": "./dist/config/snapwp-config-manager.js",
			"require": "./dist/config/snapwp-config-manager.js",
			"types": "./dist/types/config/snapwp-config-manager.d.ts"
		}
	},
	"scripts": {
		"build": "tsc --build",
		"postbuild": "npm run resolve-paths",
		"force-build": "tsc --build --force",
		"postforce-build": "npm run resolve-paths",
		"dev": "tsc-watch -b --onSuccess \"npm run resolve-paths\"",
		"resolve-paths": "resolve-tspaths",
		"lint": "eslint . --config ../../.eslintrc.json",
		"lint:fix": "eslint . --config ../../.eslintrc.json --fix"
	},
	"dependencies": {
		"@wordpress/style-engine": "^2.6.0",
		"clsx": "^2.1.1",
		"react": "^19.0.0"
	},
	"devDependencies": {
		"@types/jest": "*",
		"@types/node": "*",
		"@types/react": "*",
		"resolve-tspaths": "*",
		"tsc-watch": "*",
		"typescript": "*",
		"@snapwp/types": "*",
		"html-react-parser": "^5.1.12"
	}
}
