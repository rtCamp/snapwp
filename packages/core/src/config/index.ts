export { getConfig, setConfig } from './snapwp-config-manager';
