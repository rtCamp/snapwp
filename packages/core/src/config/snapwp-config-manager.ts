'use snapWPConfig';

import { Logger } from '@/logger';
import { generateGraphqlUrl, isValidUrl } from '@/utils';

import type { BlockDefinitions, QueryEngine } from '@snapwp/types';
import type { HTMLReactParserOptions } from 'html-react-parser';

export interface SnapWPEnv {
	/**
	 * URL prefix for WP assets loaded from 'wp-includes' dir . Defaults to `/proxy`.
	 */
	corsProxyPrefix?: string | undefined;
	/**
	 * The URL of the Next.js site. Defaults to `process.env.NEXT_PUBLIC_FRONTEND_URL`.
	 */
	frontendUrl: string;
	/**
	 * The GraphQL endpoint. Defaults to `graphql`.
	 */
	graphqlEndpoint: string;
	/**
	 * REST URL prefix. Defaults to `/wp-json`.
	 */
	restUrlPrefix: string;
	/**
	 * Uploads directory. Defaults to `/wp-content/uploads`.
	 */
	uploadsDirectory: string;
	/**
	 * The home URL of the WordPress site. Defaults to `process.env.NEXT_PUBLIC_WP_HOME_URL`.
	 */
	wpHomeUrl: string;
	/**
	 * The site URL of the WordPress site. Defaults to `process.env.NEXT_PUBLIC_WP_SITE_URL`.
	 */
	wpSiteUrl: string;
}

export interface SnapWPConfig< TClient = unknown, TClientOptions = unknown > {
	/**
	 * Block definitions for the editor.
	 */
	blockDefinitions?: BlockDefinitions;
	/**
	 * html-react-parser overload options
	 */
	parserOptions?: HTMLReactParserOptions;
	/**
	 * Query Engine
	 */
	query: QueryEngineConfig< TClient, TClientOptions >;
}

interface QueryEngineConfig< TClient, TClientOptions > {
	engine: new (
		options?: TClientOptions
	) => QueryEngine< TClient, TClientOptions >;
	options?: Parameters<
		QueryEngine< TClient, TClientOptions >[ 'init' ]
	>[ 0 ];
}

/**
 * Schema used to validate the configuration.
 */
type ConfigSchema< T > = {
	[ K in keyof T ]: {
		type: string;
		required: boolean;
		validate?: ( value: T[ K ] ) => void;
	};
};

/**
 * Default configuration.
 */
const defaultConfig: Partial< SnapWPEnv & SnapWPConfig > = {
	corsProxyPrefix:
		// eslint-disable-next-line n/no-process-env -- We're using `NODE_ENV` to derive a default value.
		process.env.NODE_ENV === 'development' ? '/proxy' : undefined,
	graphqlEndpoint: 'index.php?graphql',
	restUrlPrefix: '/wp-json',
	uploadsDirectory: '/wp-content/uploads',
};

/**
 * Get the configuration from environment variables.
 *
 * This is a function (instead of a constant) so we can inject the variables in Jest.
 *
 * @return The configuration object.
 */
// @ts-ignore - ignore check for nextUrl,homeUrl to run missing environment variable test.
const envConfig = (): Partial< SnapWPEnv > => ( {
	/* eslint-disable n/no-process-env -- These are the env variables we want to manage. */
	corsProxyPrefix: process.env.NEXT_PUBLIC_CORS_PROXY_PREFIX,
	frontendUrl: process.env.NEXT_PUBLIC_FRONTEND_URL,
	graphqlEndpoint: process.env.NEXT_PUBLIC_GRAPHQL_ENDPOINT,
	wpHomeUrl: process.env.NEXT_PUBLIC_WP_HOME_URL,
	restUrlPrefix: process.env.NEXT_PUBLIC_REST_URL_PREFIX,
	// If `wpSiteUrl` is not provided, use `wpHomeUrl`.
	wpSiteUrl:
		process.env.NEXT_PUBLIC_WP_SITE_URL ||
		process.env.NEXT_PUBLIC_WP_HOME_URL,
	uploadsDirectory: process.env.NEXT_PUBLIC_WP_UPLOADS_DIRECTORY,
	/* eslint-enable n/no-process-env -- Rule restored. */
} );

/**
 * Singleton class for managing SnapWPConfig.
 */
class SnapWPConfigManager {
	/**
	 * The configuration.
	 */
	static config: SnapWPConfig & SnapWPEnv;

	/**
	 * Flag to check if configs are set.
	 */
	static configsSet: boolean = false;

	/**
	 * The schema used to validate the configuration.
	 */
	static snapWPConfigSchema: ConfigSchema< SnapWPConfig > = {
		blockDefinitions: {
			type: 'object',
			required: false,
		},
		parserOptions: {
			type: 'object',
			required: false,
		},
		query: {
			type: 'object',
			required: true,
		},
	};

	/**
	 * The schema used to validate the configuration.
	 */
	static snapWPConfigEnvSchema: ConfigSchema< SnapWPEnv > = {
		corsProxyPrefix: {
			type: 'string',
			required: false,
		},
		frontendUrl: {
			type: 'string',
			required: true,
			/**
			 * Validate the URL.
			 *
			 * @param {string} value The value to validate.
			 *
			 * @throws {Error} If the value is invalid.
			 */
			validate: ( value ) => {
				if ( value && ! isValidUrl( value ) ) {
					throw new Error( '`frontendUrl` should be a valid URL.' );
				}
			},
		},
		graphqlEndpoint: {
			type: 'string',
			required: false,
		},
		wpHomeUrl: {
			type: 'string',
			required: true,
			/**
			 * Validate the URL.
			 *
			 * @param {string} value The value to validate.
			 *
			 * @throws {Error} If the value is invalid.
			 */
			validate: ( value ) => {
				if ( value && ! isValidUrl( value ) ) {
					throw new Error( '`wpHomeUrl` should be a valid URL.' );
				}
			},
		},
		wpSiteUrl: {
			type: 'string',
			required: false,
			/**
			 * Validate the URL.
			 *
			 * @param {string} value The value to validate.
			 * @throws {Error} If the value is invalid.
			 */
			validate: ( value ) => {
				if ( value && ! isValidUrl( value ) ) {
					throw new Error( '`wpSiteUrl` should be a valid URL.' );
				}
			},
		},
		restUrlPrefix: {
			type: 'string',
			required: false,
			/**
			 * Validate the REST URL prefix.
			 *
			 * @param {string} value The value to validate.
			 *
			 * @throws {Error} If the value is invalid.
			 */
			validate: ( value ) => {
				if ( value && ! value.startsWith( '/' ) ) {
					throw new Error(
						'`restUrlPrefix` should start with a forward slash.'
					);
				}
			},
		},
		uploadsDirectory: {
			type: 'string',
			required: false,
			/**
			 * Validate the uploads directory.
			 *
			 * @param {string} value The value to validate.
			 *
			 * @throws {Error} If the value is invalid.
			 */
			validate: ( value ) => {
				if ( value && ! value.startsWith( '/' ) ) {
					throw new Error(
						'`uploadsDirectory` should start with a forward slash.'
					);
				}
			},
		},
	};

	/**
	 * Normalizes the configuration.
	 *
	 * @param {Partial<Type>} cfg The configuration to normalize.
	 *
	 * @return The normalized configuration.
	 */
	static normalizeConfig = < T >( cfg: Partial< T > ): Partial< T > => {
		// Removing empty values.
		( Object.keys( cfg ) as Array< keyof T > ).forEach(
			( key: keyof T ) => {
				if ( cfg[ key ] === undefined ) {
					delete cfg[ key ];
				} else if (
					// @todo this should probably be moved into the schema as a sanitize callback.
					( key === 'wpHomeUrl' ||
						key === 'frontendUrl' ||
						key === 'wpSiteUrl' ) &&
					typeof cfg[ key ] === 'string'
				) {
					const value = cfg[ key ] as string;

					cfg[ key ] = ( value.endsWith( '/' )
						? value.slice( 0, -1 )
						: value ) as unknown as T[ keyof T ];
				}
			}
		);

		return cfg;
	};

	/**
	 * Get the configuration.
	 *
	 * @return The resolved configuration.
	 */
	static getConfig(): Readonly< SnapWPConfig & SnapWPEnv > {
		if ( ! SnapWPConfigManager.configsSet ) {
			// @ts-ignore -- __snapWPConfig injected from WebPack
			SnapWPConfigManager.setConfig( __snapWPConfig );
		}

		// setConfig validates the config, so we can safely assert that it is a valid configuration.
		return {
			...SnapWPConfigManager.config,
		};
	}

	/**
	 * Set the configuration.
	 *
	 * Merges the configs:
	 * 1. Environment variables.
	 * 2. Config file.
	 * 3. Default values.
	 *
	 * @param {Partial<SnapWPConfig>} cfg The configuration object.
	 */
	static setConfig( cfg?: Partial< SnapWPConfig > ): void {
		if ( SnapWPConfigManager.configsSet ) {
			Logger.error( 'Multiple calls to setConfig detected.' );
			return;
		}

		const snapWPEnv = SnapWPConfigManager.validateConfig< SnapWPEnv >(
			SnapWPConfigManager.normalizeConfig< SnapWPEnv >( envConfig() ),
			SnapWPConfigManager.snapWPConfigEnvSchema
		);

		const snapWPConfig = SnapWPConfigManager.validateConfig< SnapWPConfig >(
			SnapWPConfigManager.normalizeConfig< SnapWPConfig >( cfg || {} ),
			SnapWPConfigManager.snapWPConfigSchema
		);

		SnapWPConfigManager.config = {
			...defaultConfig,
			...snapWPEnv,
			...snapWPConfig,
		} as SnapWPConfig & Required< SnapWPEnv >;

		SnapWPConfigManager.configsSet = true;
	}

	/**
	 * Validate and resolve the configuration.
	 *
	 * @param {Partial<Type>}      config The configuration to validate.
	 * @param {ConfigSchema<Type>} schema The schema to validate the configuration against.
	 *
	 * @return The resolved configuration.
	 *
	 * @throws {Error} If the configuration is invalid.
	 */
	static validateConfig = < T >(
		config: Partial< T >,
		schema: ConfigSchema< T >
	): T => {
		if ( typeof config !== 'object' ) {
			throw new Error( 'Configs should be an object.' );
		}

		/**
		 * Validate a property.
		 *
		 * @param {keyof Type} key The property key.
		 *
		 * @throws {Error} If the property is invalid.
		 */
		const validateProperty = ( key: keyof T ): void => {
			const prop = schema[ key ];
			const value = config[ key ];

			if ( ! prop ) {
				throw new Error( `Unknown property: ${ String( key ) }.` );
			}

			if ( prop.required && ! value ) {
				throw new Error(
					`Missing required property: ${ String( key ) }.`
				);
			}

			if ( value && prop.type && typeof value !== prop.type ) {
				throw new Error(
					`Property ${ String( key ) } should be of type ${
						prop.type
					}.`
				);
			}

			if ( prop.validate ) {
				//@ts-ignore
				prop.validate( value );
			}
		};

		// Validate each property.
		( Object.keys( config ) as Array< keyof T > ).forEach(
			( key: keyof T ) => {
				validateProperty( key );
			}
		);

		// If all properties are valid, return the resolved configuration.
		return config as T;
	};

	/**
	 * Get the GraphQL URL.
	 *
	 * @return The GraphQL URL.
	 */
	static getGraphqlUrl(): string {
		return generateGraphqlUrl(
			SnapWPConfigManager.getConfig().wpSiteUrl,
			SnapWPConfigManager.getConfig().graphqlEndpoint
		);
	}
}

export const getConfig = SnapWPConfigManager.getConfig;
export const setConfig = SnapWPConfigManager.setConfig;
export const getGraphqlUrl = SnapWPConfigManager.getGraphqlUrl;

// Exporting for testing purposes.
export const _private =
	// eslint-disable-next-line n/no-process-env -- `NODE_ENV` determines the environment.
	process.env.NODE_ENV === 'test'
		? {
				SnapWPConfigManager,
		  }
		: {};
