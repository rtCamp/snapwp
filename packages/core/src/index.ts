export * from './utils';
export * from './logger';
export * from './errors';
export * from './props';
export * from './query-engine-base';
