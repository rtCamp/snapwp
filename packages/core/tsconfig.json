{
	"extends": "../../config/tsconfig.base.json",
	"compilerOptions": {
		"rootDir": "src",
		"declarationDir": "./dist/types",
		"outDir": "./dist",
		"baseUrl": ".",
		"paths": {
			"@/*": [ "./src/*" ]
		},
		"types": [ "jest", "node" ],
		"typeRoots": [ "../../node_modules/@types" ]
	},
	"include": [ "src/**/*", "../types/dist/types/global/env.d.ts" ],
	"exclude": [ "node_modules", "dist" ],
	"references": [ { "path": "../types" } ]
}
