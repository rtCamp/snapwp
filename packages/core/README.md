# @snapwp/core

Core utilities and shared functionality powering SnapWP's composeable framework for Headless WordPress.

[Installation](#installation) | [Usage](#usage) | [Contributing](#contributing) | [rtCamp](https://rtcamp.com/)

## Installation

```bash
npm install @snapwp/core
```

## Usage

### Utils

```javascript
import { findElementAndGetClassNames } from '@snapwp/core';

const renderedHTML =
	'<div class="class1"><div class="class2" id="div-2"></div></div>';
const classNames = findElementAndGetClassNames( renderedHTML, '#div-2' ); // class2
```

```javascript
import { generateGraphqlUrl } from '@snapwp/core';

const url = generateGraphqlUrl( homeUrl, graphqlEndpoint );
```

```javascript
import { getClassNamesFromString } from '@snapwp/core';

const htmlString = '<div class="class1 class2 class3"></div>';
const classNames = getClassNamesFromString( htmlString ); // [ class1, class2, class3 ]
```

```javascript
import { getColorClassName } from '@snapwp/core';

const colorClassName = getColorClassName( 'red', 'background' ); // has-red-background
```

### Config Manager

```javascript
import { getConfig } from '@snapwp/core/config';

const config = getConfig();
```

## Contributing

This package is part of a monorepo for SnapWP and is actively maintained by [rtCamp](https://rtcamp.com/). Packages are published to [npm](https://www.npmjs.com/) from the `packages` directory and can be used standalone in the headless WordPress ecosystem or as part of SnapWP's framework.

Contributions are _welcome_ and **encouraged!** To learn more about contributing to this package or SnapWP as a whole, please read the [Contributing Guide](../../../.github/CONTRIBUTING.md).

For development guidelines, please refer to our [Development Guide](../../DEVELOPMENT.md).

## Want to expand what's possible with WordPress?

<a href="https://rtcamp.com/"><img src="https://rtcamp.com/wp-content/uploads/sites/2/2019/04/github-banner@2x.png" alt="Join us at rtCamp, we specialize in providing high performance enterprise WordPress solutions."></a>
