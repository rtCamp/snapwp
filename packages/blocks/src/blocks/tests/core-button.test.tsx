import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import CoreButton from '../core-button';
import { getConfig } from '@snapwp/core/config';

jest.mock( '@wordpress/style-engine', () => {
	const actual = jest.requireActual( '@wordpress/style-engine' );
	return {
		...actual,
		compileCSS: jest.fn( ( style ) => style ),
	};
} );

describe( 'CoreButton Component', () => {
	let ORIG_ENV: NodeJS.ProcessEnv;

	beforeEach( () => {
		// @ts-ignore Allow setting global variable for testing
		ORIG_ENV = { ...process.env };
		// @ts-ignore Allow emptying global variable for testing
		global.backupSnapWPConfig = { ...global.__snapWPConfig };
	} );

	afterEach( () => {
		// @ts-ignore Allow emptying global variable for testing
		global.__snapWPConfig = global.backupSnapWPConfig;
		// @ts-ignore Allow emptying global variable for testing
		delete global.backupSnapWPConfig;
		process.env = {
			...ORIG_ENV,
		};
	} );

	test( 'renders with correct tag and attributes for an anchor tag', () => {
		const exampleAttributes = {
			cssClassName: 'test-class',
			linkClassName: 'link-class',
			style: JSON.stringify( { color: 'red' } ),
			text: 'Click Me',
			title: 'Test Button',
			url: 'https://example.com',
			tagName: 'a',
			rel: 'noopener',
			linkTarget: '_blank',
			buttonType: 'button',
		};

		const { asFragment } = render(
			<CoreButton attributes={ exampleAttributes } />
		);
		const linkElement = screen.getByRole( 'link', { name: 'Click Me' } );

		expect( linkElement ).toBeInTheDocument();
		expect( linkElement ).toHaveAttribute( 'href', 'https://example.com' );
		expect( linkElement ).toHaveAttribute( 'target', '_blank' );
		expect( linkElement ).toHaveAttribute( 'rel', 'noopener' );
		expect( linkElement ).toHaveClass( 'link-class' );
		expect( linkElement ).toHaveStyle( 'color: red' );
		expect( asFragment() ).toMatchSnapshot();
	} );

	test( 'renders with corrected url', () => {
		const { homeUrl } = getConfig();
		const exampleAttributes = {
			cssClassName: 'test-class',
			linkClassName: 'link-class',
			style: JSON.stringify( { color: 'red' } ),
			text: 'Click Me',
			title: 'Test Button',
			url: homeUrl + '/sub',
			tagName: 'a',
			rel: 'noopener',
			linkTarget: '_blank',
			buttonType: 'button',
		};

		const { asFragment } = render(
			<CoreButton attributes={ exampleAttributes } />
		);
		const linkElement = screen.getByRole( 'link', { name: 'Click Me' } );

		expect( linkElement ).toHaveAttribute( 'href', '/sub' );
		expect( asFragment() ).toMatchSnapshot();
	} );

	test( 'renders with correct tag and attributes for a button tag', () => {
		const buttonAttributes = {
			cssClassName: 'test-class',
			linkClassName: 'link-class',
			style: JSON.stringify( { color: 'red' } ),
			text: 'Click Me',
			title: 'Test Button',
			tagName: 'button',
			buttonType: 'submit',
		};
		const { asFragment } = render(
			<CoreButton attributes={ buttonAttributes } />
		);
		const buttonElement = screen.getByRole( 'button', {
			name: 'Click Me',
		} );

		expect( buttonElement ).toBeInTheDocument();
		expect( buttonElement ).toHaveAttribute( 'type', 'submit' );
		expect( buttonElement ).not.toHaveAttribute( 'href' );
		expect( buttonElement ).not.toHaveAttribute( 'target' );
		expect( buttonElement ).toHaveClass( 'link-class' );
		expect( buttonElement ).toHaveStyle( 'color: red' );
		expect( asFragment() ).toMatchSnapshot();
	} );

	test( 'applies default tag "a" when no tagName is provided', () => {
		const attributesWithoutTag = {
			cssClassName: 'test-class',
			linkClassName: 'link-class',
			style: JSON.stringify( { color: 'red' } ),
			text: 'Click Me',
			title: 'Test Button',
			url: 'https://example.com',
			rel: 'noopener',
			linkTarget: '_blank',
			buttonType: 'button',
		};
		const { asFragment } = render(
			<CoreButton attributes={ attributesWithoutTag } />
		);
		const linkElement = screen.getByRole( 'link', { name: 'Click Me' } );

		expect( linkElement.tagName ).toBe( 'A' );
		expect( linkElement ).toHaveAttribute( 'href', 'https://example.com' );
		expect( linkElement ).toHaveStyle( 'color: red' );
		expect( asFragment() ).toMatchSnapshot();
	} );

	test( 'renders without text if no text is provided', () => {
		const attributesWithoutText = {
			cssClassName: 'test-class',
			linkClassName: 'link-class',
			style: JSON.stringify( { color: 'red' } ),
			title: 'Test Button',
			url: 'https://example.com',
			tagName: 'a',
			rel: 'noopener',
			linkTarget: '_blank',
			buttonType: 'button',
		};
		const { asFragment } = render(
			<CoreButton attributes={ attributesWithoutText } />
		);
		const linkElement = screen.getByRole( 'link' );

		expect( linkElement ).toBeEmptyDOMElement();
		expect( linkElement ).toHaveStyle( 'color: red' );
		expect( asFragment() ).toMatchSnapshot();
	} );

	test( 'renders without optional attributes if not provided', () => {
		const minimalAttributes = {
			text: 'Minimal Button',
		};
		const { asFragment, container } = render(
			<CoreButton attributes={ minimalAttributes } />
		);
		const linkElement = container.querySelector( 'a' );

		expect( linkElement ).not.toBeNull();
		expect( linkElement ).not.toHaveAttribute( 'href' );
		expect( linkElement ).not.toHaveAttribute( 'target' );
		expect( linkElement ).not.toHaveAttribute( 'rel' );
		expect( linkElement ).not.toHaveClass();
		expect( asFragment() ).toMatchSnapshot();
	} );

	test( 'renders correctly with custom CSS classes', () => {
		const customClassAttributes = {
			cssClassName: 'custom-class',
			linkClassName: 'link-class',
			style: JSON.stringify( { color: 'blue' } ),
			text: 'Custom Button',
			title: 'Custom Test Button',
			url: 'https://example.com',
			tagName: 'a',
			rel: 'noopener',
			linkTarget: '_blank',
		};
		const { asFragment } = render(
			<CoreButton attributes={ customClassAttributes } />
		);
		const linkElement = screen.getByRole( 'link', {
			name: 'Custom Button',
		} );

		expect( linkElement ).toHaveClass( 'link-class' );
		expect( linkElement ).toHaveStyle( 'color: blue' );
		expect( asFragment() ).toMatchSnapshot();
	} );

	test( 'renders with empty attributes object', () => {
		const { asFragment, container } = render(
			<CoreButton attributes={ {} } />
		);
		const linkElement = container.querySelector( 'a' );

		expect( linkElement ).not.toBeNull();
		expect( linkElement!.tagName ).toBe( 'A' );
		expect( linkElement ).toBeEmptyDOMElement();
		expect( asFragment() ).toMatchSnapshot();
	} );

	test( 'renders with undefined attributes', () => {
		const { asFragment, container } = render(
			<CoreButton attributes={ undefined } />
		);
		const linkElement = container.querySelector( 'a' );

		expect( linkElement ).not.toBeNull();
		expect( linkElement!.tagName ).toBe( 'A' );
		expect( linkElement ).toBeEmptyDOMElement();
		expect( asFragment() ).toMatchSnapshot();
	} );
} );
