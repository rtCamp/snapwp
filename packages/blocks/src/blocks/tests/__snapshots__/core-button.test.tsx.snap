// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CoreButton Component applies default tag "a" when no tagName is provided 1`] = `
<DocumentFragment>
  <div
    class="test-class"
  >
    <a
      class="link-class"
      href="https://example.com"
      rel="noopener"
      style="color: red;"
      target="_blank"
      title="Test Button"
    >
      Click Me
    </a>
  </div>
</DocumentFragment>
`;

exports[`CoreButton Component renders correctly with custom CSS classes 1`] = `
<DocumentFragment>
  <div
    class="custom-class"
  >
    <a
      class="link-class"
      href="https://example.com"
      rel="noopener"
      style="color: blue;"
      target="_blank"
      title="Custom Test Button"
    >
      Custom Button
    </a>
  </div>
</DocumentFragment>
`;

exports[`CoreButton Component renders with correct tag and attributes for a button tag 1`] = `
<DocumentFragment>
  <div
    class="test-class"
  >
    <button
      class="link-class"
      style="color: red;"
      title="Test Button"
      type="submit"
    >
      Click Me
    </button>
  </div>
</DocumentFragment>
`;

exports[`CoreButton Component renders with correct tag and attributes for an anchor tag 1`] = `
<DocumentFragment>
  <div
    class="test-class"
  >
    <a
      class="link-class"
      href="https://example.com"
      rel="noopener"
      style="color: red;"
      target="_blank"
      title="Test Button"
    >
      Click Me
    </a>
  </div>
</DocumentFragment>
`;

exports[`CoreButton Component renders with corrected url 1`] = `
<DocumentFragment>
  <div
    class="test-class"
  >
    <a
      class="link-class"
      href="https://env-next.example.com/sub"
      rel="noopener"
      style="color: red;"
      target="_blank"
      title="Test Button"
    >
      Click Me
    </a>
  </div>
</DocumentFragment>
`;

exports[`CoreButton Component renders with empty attributes object 1`] = `
<DocumentFragment>
  <div
    class=""
  >
    <a />
  </div>
</DocumentFragment>
`;

exports[`CoreButton Component renders with null attributes 1`] = `
<DocumentFragment>
  <div
    class=""
  >
    <a />
  </div>
</DocumentFragment>
`;

exports[`CoreButton Component renders with undefined attributes 1`] = `
<DocumentFragment>
  <div
    class=""
  >
    <a />
  </div>
</DocumentFragment>
`;

exports[`CoreButton Component renders without optional attributes if not provided 1`] = `
<DocumentFragment>
  <div
    class=""
  >
    <a>
      Minimal Button
    </a>
  </div>
</DocumentFragment>
`;

exports[`CoreButton Component renders without text if no text is provided 1`] = `
<DocumentFragment>
  <div
    class="test-class"
  >
    <a
      class="link-class"
      href="https://example.com"
      rel="noopener"
      style="color: red;"
      target="_blank"
      title="Test Button"
    />
  </div>
</DocumentFragment>
`;
